### ISSUE TITLE
blocker(model): #12641 Flash 0731 oracle cannot run - no native Flash 0731 Prefill/Step seam exists yet

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12935]
﻿

# blocker(model): #12641 Flash 0731 oracle cannot run - no native Flash 0731 Prefill/Step seam exists yet

```routing
lane: model
paths: ["internal/model/v4_flash_attention.go", "internal/model/v4_flash_kv_layout.go", "internal/model/generate_context.go", "internal/model/config.go"]
expected_steps: 2
public_issue: anthony-chaudhary/fak#12641
```

## Current state

[fak#12641](https://github.com/anthony-chaudhary/fak/issues/12641) asks for an
independent CPU numeric oracle for the pinned DeepSeek V4 Flash 0731 token path:
compare production `Session.Prefill`/`Session.Step` logits against an
independently derived scalar reference over mHC four-stream mixing, ratio
0/4/128 attention, Flash routing, FP4/FP8 decode, and final head collapse.

That oracle cannot be built today: there is no native Flash 0731 token forward
to oracle against. `v4_flash_attention.go` contains no attention math, the
ratio-0 "tracer" runs the generic Llama-shape GQA path, and every ratio-4/128
layer fails closed before prefill. This is the same class of blocker as
[#12913](https://github.com/anthony-chaudhary/fak/issues/12913) for the V4.1
oracle (#12900).

Evidence at revision `7872f01b1d1fe23eabc4c98b48bffcef5a386062`, checked 2026-09-12:

1. The only fail-closed guard, `refuseUnimplementedV4FlashAttention`
   (`internal/model/v4_flash_attention.go:99`), is called from exactly one
   place, `GenerateContext` (`internal/model/generate_context.go:20`).
   `Prefill`, `Step`, `Generate`, and `Forward` never call it.
2. `deepseek_v4` does not satisfy `usesMLAMoELayout()`
   (`internal/model/config.go:1080`), so a ratio-0 session falls through to the
   generic Q/K/V path in `forward.go` and never touches
   `V4FlashCircularWindow`, a compressor, or an indexer.
3. `newV4FlashKVState` (`internal/model/v4_flash_kv_layout.go:106`) has zero
   production call sites; compressor/indexer bodies for ratios 4/128 are
   explicitly not implemented (`v4_flash_kv_layout.go:23-27`).
4. Real Flash geometry (`compress_ratios: [0,0,4,128,...]`, 46 entries) is
   dominated by 4/128, so `GenerateContext` returns
   `ErrV4FlashAttentionUnimplemented` before any token is produced.
5. mHC pre/post four-stream mixing, Sinkhorn, combination-matrix orientation,
   grouped output projection, and head collapse have no production
   implementation; `hc_mult`/`hc_eps`/`hc_sinkhorn_iters` are admitted metadata
   only (`v4_config_admit.go:53-55`).

## Problem frame

- Centrality: Core: the independent oracle gates every Flash forward claim.
- P1: advanced - an oracle cannot be built against an absent forward seam.
- P2: advanced - prevents dispatching a leaf into a wall.
- P3: preserved - existing component oracles stay valid.
- P4: advanced - routes the unblock to the exact dependency (#12636).

## Parent context

Parent track: #12635 (native DeepSeek V4 Flash 0731 inference).

## Why now

The #12641 oracle leaf is dispatchable-looking but cannot run; recording the
blocker prevents wasted dispatch and routes the unblock to #12636.

## Scope

In scope: record this blocker, keep the #12641 oracle leaf parked behind it, and
route the unblock to #12636 (dedicated V4 session attention path). The #12641
oracle scope items that already have a production seam are already independently
oracled - routing (`v4_router_test.go:9`, `v4_flash_hash_test.go:13`), expert
compose (`v4_expert_runtime_test.go:177`), FP4 decode (`v4_flash_quant_test.go:21`),
FP8 dense decode (`v4_flash_dense_quant_test.go`) - so the residual gap is
entirely the absent forward.

Out of scope: implementing the V4 attention, KV, MoE, or quant forward leaves
themselves; those land under #12636-#12639.

## Core through-line

Turn the silent absence of a native Flash 0731 token forward into a witnessed,
routed blocker so the #12641 oracle leaf is not dispatched into a wall.

## Working spine

Obstacle intake (#12641 oracle leaf) -> evidence capture at pinned revision ->
witnessed blocker + dependency routing to #12636 -> #12641 oracle resumes once the
native Flash 0731 Prefill/Step seam lands.

## Gold-plating boundary

No new model code, no fixture that tests only itself, no V4.1 labeling, and no
simultaneous CSA/HCA encoding. This issue only records evidence and routes it.

## Dependencies and scope fences

- Blocking: [#12636](https://github.com/anthony-chaudhary/fak/issues/12636) V4
  session attention path (ratio 0 first, then 4 and 128).
- Transitively blocking: [#12637](https://github.com/anthony-chaudhary/fak/issues/12637)
  KV state, [#12638](https://github.com/anthony-chaudhary/fak/issues/12638)
  expert execution, [#12639](https://github.com/anthony-chaudhary/fak/issues/12639)
  FP8/FP4 execution.
- Blocked-by record for: [#12641](https://github.com/anthony-chaudhary/fak/issues/12641).

## Witness

```sh
go -C ..\fak test ./internal/model -run 'TestV4Flash' -count=1
```

This passes today and witnesses the fail-closed boundary plus the generic-path
fallback. It does not, and cannot yet, witness Flash 0731 token numerics.

### File:Line Seam
`internal/model/v4_flash_attention.go:99`

### Blast Radius & Affected Lanes
- **Primary Lane:** `model`
- **Unaffected Lanes:** no production code changes; trunk remains green

### Quarantined Fallback Mechanism
The #12641 oracle leaf stays parked behind this blocker; existing V4 component
oracles continue to cover routed-expert routing, quant decode, and regime
classification. No generic path is silently substituted for compressed
attention - the forward fails closed.

## Scoped Acceptance Criteria

- [ ] 1. Blocker evidence recorded and cross-linked to #12641 and #12636.
- [ ] 2. #12641 oracle leaf marked blocked pending the native Prefill/Step seam.
- [ ] 3. Witness command above passes on trunk.
- [ ] 4. Blocker closes when #12636 lands the ratio-0 native V4 session path.

## Done condition

The blocker is open with the `quarantine` label and the evidence above; #12641
carries a blocked-by reference to it; the witness command passes. The blocker is
resolved only when a native `Session.Prefill`/`Session.Step` seam executes the
Flash 0731 attention path, enabling the #12641 oracle.

## Definition of done

Blocker is filed, routed, and linked; #12641 reflects the dependency; the
witness command is green.

## Acceptance gate

The witness command passes on trunk and the blocker is linked from #12641.

## Closure binding

This blocker closes when #12636 lands a native ratio-0 V4 Flash session
attention path enabling the #12641 oracle; the resolving commit cites #12641.

## Work estimate

Estimate: 2 points (bounded blocker record).

## Overall completion contribution

Contribution: 2/40 points (blocker routing and evidence record).

## Completion standard

development

## Ref

- Parent track: [#12635](https://github.com/anthony-chaudhary/fak/issues/12635)
- Blocked oracle: [#12641](https://github.com/anthony-chaudhary/fak/issues/12641)
- Precedent: [#12913](https://github.com/anthony-chaudhary/fak/issues/12913)
